state/watcher.go

```diff
@@ -496,6 +497,185 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachineStartTimes watches the non-container machines in the model
+// for changes to the Life or AgentStartTime fields and reports them as a batch
+// after the specified quiesceInterval time has passed without seeing any new
+// change events.
+func (st *State) WatchModelMachineStartTimes(quiesceInterval time.Duration) StringsWatcher {
+	return newModelMachineStartTimeWatcher(st, st.clock(), quiesceInterval)
+}
+
+type modelMachineStartTimeFieldDoc struct {
+	Id             string    `bson:"_id"`
+	Life           Life      `bson:"life"`
+	AgentStartedAt time.Time `bson:"agent-started-at"`
+}
+
+type modelMachineStartTimeWatcher struct {
+	commonWatcher
+	outCh chan []string
+
+	clk             clock.Clock
+	quiesceInterval time.Duration
+	seenDocs        map[string]modelMachineStartTimeFieldDoc
+}
+
+func newModelMachineStartTimeWatcher(backend modelBackend, clk clock.Clock, quiesceInterval time.Duration) StringsWatcher {
+	w := &modelMachineStartTimeWatcher{
+		commonWatcher:   newCommonWatcher(backend),
+		outCh:           make(chan []string),
+		clk:             clk,
+		quiesceInterval: quiesceInterval,
+		seenDocs:        make(map[string]modelMachineStartTimeFieldDoc),
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.outCh)
+		return w.loop()
+	})
+	return w
+}
+
+// Changes returns the event channel for the watcher.
+func (w *modelMachineStartTimeWatcher) Changes() <-chan []string {
+	return w.outCh
+}
+
+func (w *modelMachineStartTimeWatcher) loop() error {
+	docWatcher := newCollectionWatcher(w.backend, colWCfg{col: machinesC})
+	defer func() { _ = docWatcher.Stop() }()
+
+	var (
+		timer           = w.clk.NewTimer(w.quiesceInterval)
+		timerArmed      = true
+		unprocessedDocs = make(set.Strings)
+		outCh           chan []string
+		changeSet       []string
+	)
+	defer func() { _ = timer.Stop() }()
+
+	// Collect and initial set of machine IDs; this makes the worker
+	// compatible with other workers that expect the full state to be
+	// immediately emitted once the worker starts.
+	initialSet, err := w.initialMachineSet()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	changeSet = initialSet.Values()
+	outCh = w.outCh
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case changes := <-docWatcher.Changes():
+			if len(changes) == 0 {
+				continue
+			}
+			for _, docID := range changes {
+				unprocessedDocs.Add(w.backend.docID(docID))
+			}
+
+			// Restart the timer if currently stopped.
+			if !timerArmed {
+				_ = timer.Reset(w.quiesceInterval)
```

Comment:
I think you need to set timerArmed here, otherwise you'll still be pushing the timer out on each subsequent event, right?

Reply:
Doh! I am pretty sure I had that in there and it got lost while rebasing... Nice catch! I will push a commit to add it back in!

Reply:
Fixed via https://github.com/juju/juju/pull/10902/commits/1f0026be0ebbfbddd177cde08ac267fd117e8c6d